src/PsrLogAdapter.php

```diff
@@ -0,0 +1,118 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB;
+
+use MongoDB\Driver\Monitoring\LogSubscriber;
+use MongoDB\Exception\UnexpectedValueException;
+use Psr\Log\LoggerInterface;
+use Psr\Log\LogLevel;
+use SplObjectStorage;
+
+use function MongoDB\Driver\Monitoring\addSubscriber;
+use function MongoDB\Driver\Monitoring\removeSubscriber;
+use function sprintf;
+
+final class PsrLogAdapter implements LogSubscriber
+{
+    private static ?self $instance = null;
+
+    /** @psalm-var SplObjectStorage<LoggerInterface, null> */
+    private SplObjectStorage $loggers;
+
+    private const PSR_LEVELS = [
+        LogSubscriber::LEVEL_ERROR => LogLevel::ERROR,
+        /* libmongoc considers "critical" less severe than "error" so map it to
+         * "error" in the PSR logger. */
+        LogSubscriber::LEVEL_CRITICAL => LogLevel::ERROR,
+        LogSubscriber::LEVEL_WARNING => LogLevel::WARNING,
+        LogSubscriber::LEVEL_MESSAGE => LogLevel::NOTICE,
+        LogSubscriber::LEVEL_INFO => LogLevel::INFO,
+        LogSubscriber::LEVEL_DEBUG => LogLevel::DEBUG,
+    ];
+
+    public static function addLogger(LoggerInterface $logger): void
+    {
+        $instance = self::getInstance();
+
+        $instance->loggers->attach($logger);
+
+        addSubscriber($instance);
+    }
+
+    /**
+     * Pipes a log message from PHPC to all registered PSR loggers.
+     *
+     * @internal
+     * @see LogSubscriber::log()
+     */
+    public function log(int $level, string $domain, string $message): void
+    {
+        if (! isset(self::PSR_LEVELS[$level])) {
+            throw new UnexpectedValueException(sprintf(
+                'Expected level to be >= %d and <= %d, %d given for domain "%s" and message: %s',
+                LogSubscriber::LEVEL_ERROR,
+                LogSubscriber::LEVEL_DEBUG,
+                $level,
+                $domain,
+                $message,
+            ));
+        }
+
+        $instance = self::getInstance();
+        $psrLevel = self::PSR_LEVELS[$level];
+        $context = ['domain' => $domain];
+
+        foreach ($instance->loggers as $logger) {
+            $logger->log($psrLevel, $message, $context);
+        }
+    }
+
+    public static function removeLogger(LoggerInterface $logger): void
+    {
+        $instance = self::getInstance();
+        $instance->loggers->detach($logger);
+
+        if ($instance->loggers->count() === 0) {
+            removeSubscriber($instance);
+        }
+    }
+
+    /**
+     * Writes a log message to all registered PSR loggers.
+     *
+     * @internal
+     */
+    public static function writeLog(int $level, string $domain, string $message): void
```

Comment:
Per https://github.com/mongodb/mongo-php-driver/pull/1395#discussion_r1326793685, PHPC defines a `mongoc_log()` function for internal tests. I'd ultimately like to have it defined only for test builds ([PHPC-2293](https://jira.mongodb.org/browse/PHPC-2293)), so I think it still makes sense to use our own `writeLog()` function here.

We also discussed only caring about logging through PsrLogAdapter for spec compliance, so it shouldn't be necessary to have PHPLIB log directly to a PHPC mechanism (be it the `mongodb.debug` stream or registered LogSubscriber object(s)).

Reply:
Since we're only dealing with PSR loggers here, does it make any sense to use the LogSubscriber levels here? The [PSR levels](https://www.php-fig.org/psr/psr-3/#5-psrlogloglevel) are consistent with our [logging spec](https://github.com/mongodb/specifications/blob/master/source/logging/logging.rst#log-severity-levels), although the spec defines an additional trace level (how ironic). That said, it's not used so I don't think there's any urgent need to support it.

If we use PSR levels, we'd need a non-optional dependency on `psr/log`. It may make more sense to define our own levels (exactly consistent with the spec) and then maintain a mapping. This seems to be what Monolog does internally (for both PSR and RFC5424).

Reply:
I think people will get confused either way. Since PSR log levels are strings and libmongoc log levels are integers, how about accepting both?

Reply:
@GromNaN is correct that no users should be calling `writeLog()` directly. It is only intended for internal use within PHPLIB.

That said, after reading through the [Logging spec](https://github.com/mongodb/specifications/blob/master/source/logging/logging.rst#log-severity-levels) earlier today, I realized we should probably use those levels directly. My most recent commit introduces constants for the spec levels and changes `writeLog()` to take those and translate them internally to a PSR level (the only non-standard mapping is spec level `trace` to PSR level `debug`).

On a related note, I also changed `writeLog()` to call the PSR loggers directly instead of forwarding to `LogSubscriber::log()`, since forwarding would limit what levels PHPLIB could utilize (probably not relevant, but worth noting).